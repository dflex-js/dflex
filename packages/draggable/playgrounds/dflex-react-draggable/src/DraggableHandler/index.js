import DraggableHandler from "./DraggableHandler";

export default DraggableHandler;
