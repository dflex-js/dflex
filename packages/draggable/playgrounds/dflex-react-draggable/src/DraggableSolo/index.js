import DraggableSolo from "./DraggableSolo";

export default DraggableSolo;
