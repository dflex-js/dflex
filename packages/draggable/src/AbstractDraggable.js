import { DRAGGED_ELM } from "../constants.json";

/**
 * Style that will be added to dragged element.
 *
 * @static
 * @property draggedStyleProps,
 * @type {Array}
 * @readonly
 * @memberof AbstractDraggable
 */
const draggedStyleProps = [
  {
    prop: "pointerEvents",
    value: "none",
  },
  {
    prop: "zIndex",
    value: "1",
  },
  {
    prop: "transition",
    value: `opacity 0.2s cubic-bezier(0.2, 0, 0, 1) 0s`,
  },
];

/**
 * AbstractDraggable element.
 *
 * @class AbstractDraggable
 */
class AbstractDraggable {
  /**
   * Creates an instance of AbstractDraggable.
   * Works Only on dragged element level.
   *
   * @param {Object}  element - object reference from store
   * @param {Object}  initCoordinates
   * @param {number}  initCoordinates.x
   * @param {number}  initCoordinates.y
   *
   * @memberof AbstractDraggable
   */
  constructor(element, { x: initX, y: initY }) {
    /**
     * Assign instance for dragged.
     */
    this[DRAGGED_ELM] = element;

    const {
      translateX,
      translateY,
      element: { style: draggedStyle },
    } = this[DRAGGED_ELM];

    this.draggedStyle = draggedStyle;

    /**
     * When dragging start, element shouldn't jump from its translate. So, we
     * calculate offset that make translate X,Y start from zero:
     *  goToX = x + this.outerOffsetX.
     *  goToY = y + this.outerOffsetY.
     *
     * goToX and goToY both should be zero with first click. Starts with simple
     * equating: initX = X. Taking into considerations translate value.
     *
     */
    this.outerOffsetX = -initX + translateX;
    this.outerOffsetY = -initY + translateY;

    this.setDragged(true);
  }

  /**
   * Triggers twice. Once when constructor is initiated, the other when drag is
   * ended. It adds/removes style.
   *
   * @param {boolean} isActive - is dragged operation active or it is ended.
   * @memberof AbstractDraggable
   */
  setDragged(isActive) {
    if (isActive) {
      draggedStyleProps.forEach(({ prop, value }) => {
        this.draggedStyle[prop] = value;
      });

      return;
    }

    /**
     * Not active: end of dragging.
     */
    this.draggedStyle.pointerEvents = null;
  }

  endDragging() {
    this.setDragged(false);
  }

  /**
   * Executes dragging by applying transform.
   * Writes to draggedElmCurrentOffset in Transform class.
   * Set values to isDragged flags.
   *
   * @param {number} x - mouse x coordinates
   * @param {number} y - mouse y coordinates
   * @memberof AbstractDraggable
   */
  dragAt(x, y) {
    /**
     * Calculates translate coordinates.
     *
     * Indicates dragged y-transformation that's will be updated during the
     * dropping process. Updating Y immediately will effect calculations in
     * transform, that's why it is updated when dragging is done.
     */
    this[DRAGGED_ELM].translateX = x + this.outerOffsetX;
    this[DRAGGED_ELM].translateY = y + this.outerOffsetY;

    this.draggedStyle.transform = `translate(${this[DRAGGED_ELM].translateX}px,${this[DRAGGED_ELM].translateY}px)`;
  }
}

export default AbstractDraggable;
