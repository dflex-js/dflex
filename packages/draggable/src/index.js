import Draggable from "./Draggable";
import store from "./DraggableStoreImp";

export { Draggable, store };
