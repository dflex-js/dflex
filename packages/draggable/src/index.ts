export { default as Draggable } from "./Draggable";
export { default as store } from "./DraggableStoreImp";
export { default as AbstractDraggable } from "./AbstractDraggable";

export type { DraggedStyle, AbstractDraggableInterface } from "./types";
