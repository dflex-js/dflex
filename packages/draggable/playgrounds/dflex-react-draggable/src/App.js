import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";

import Lists from "./Lists";
import DraggableSolo from "./DraggableSolo";

function App() {
  return (
    <Router>
      <Switch>
        <Route path="/draggable-solo">
          <DraggableSolo />
        </Route>

        <Route path="/">
          <Lists />
        </Route>
      </Switch>
    </Router>
  );
}

export default App;
