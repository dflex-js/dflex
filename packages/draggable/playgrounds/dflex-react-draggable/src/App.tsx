import React from "react";
import { BrowserRouter, Routes, Route } from "react-router-dom";

import DraggableSolo from "./DraggableSolo";
import DraggableHandler from "./DraggableHandler";

function App() {
  return (
    <BrowserRouter>
      <Routes>
        <Route path="/draggable-handler" element={<DraggableHandler />}></Route>
        <Route path="/" element={<DraggableSolo />}></Route>
      </Routes>
    </BrowserRouter>
  );
}

export default App;
