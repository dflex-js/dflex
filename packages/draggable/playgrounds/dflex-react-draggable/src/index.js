/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from "react";
import ReactDOM from "react-dom";
import "./index.css";
import App from "./App";

ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById("root")
);
