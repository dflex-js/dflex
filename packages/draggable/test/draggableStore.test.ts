import { store } from "../src";

const ref = document.createElement("div");

const elm0D0 = {
  id: "id-0",
  depth: 0,
  ref,
};

describe("Draggable Store", () => {
  beforeAll(() => {
    store.register(elm0D0);
  });

  it("Registry is not empty", () => {
    // Why this is an error? Do you know why? If so, there's a PR waiting for you.
    // @ts-expect-error
    expect(store.registry).toBeTruthy();
  });

  it("Registers element and initiates translateX,Y", () => {
    // @ts-expect-error
    expect(store.registry).toMatchSnapshot();
  });
});
