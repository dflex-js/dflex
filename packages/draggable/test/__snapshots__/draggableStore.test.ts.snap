// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draggable Store Registers element and initiates translateX,Y 1`] = `
Object {
  "id-0": CoreInstance {
    "animatedFrame": null,
    "currentLeft": NaN,
    "currentTop": NaN,
    "hasToTransform": false,
    "id": "id-0",
    "isInitialized": true,
    "isPaused": false,
    "isVisible": undefined,
    "keys": Object {
      "chK": null,
      "pK": "1-0",
      "sK": "0-0",
    },
    "offset": Object {
      "height": 0,
      "left": NaN,
      "top": NaN,
      "width": 0,
    },
    "order": Object {
      "parent": 0,
      "self": 0,
    },
    "prevTranslateY": Array [],
    "ref": <div />,
    "translateX": 0,
    "translateY": 0,
  },
}
`;
